File: cloudinary_to_catalog.js
const fs = require("fs");
const path = require("path");
const axios = require("axios");
require("dotenv").config();

const CLOUD_NAME = process.env.CLOUDINARY_CLOUD_NAME;
const API_KEY = process.env.CLOUDINARY_API_KEY;
const API_SECRET = process.env.CLOUDINARY_API_SECRET;
const BASE_FOLDER = path.join(__dirname, "fabric-images");
const OUTPUT_FILE = path.join(__dirname, "data", "catalog.json");
const SKIP_IDS = new Set(["REVIEW_REQUIRED"]);

function getQualities() {
  return fs
    .readdirSync(BASE_FOLDER, { withFileTypes: true })
    .filter((entry) => entry.isDirectory())
    .map((entry) => entry.name)
    .sort();
}

function extractDesignId(publicId) {
  const filename = publicId.split("/").pop().toUpperCase();
  const parts = filename.split("_");

  if (parts.length >= 3) {
    return parts.slice(0, -1).join("_");
  }

  return filename;
}

async function fetchQualityCatalog(quality) {
  const url = `https://api.cloudinary.com/v1_1/${CLOUD_NAME}/resources/search`;
  const response = await axios.post(
    url,
    {
      expression: `folder=fabric-images/${quality}`,
      max_results: 500,
      sort_by: [{ public_id: "asc" }]
    },
    {
      auth: {
        username: API_KEY,
        password: API_SECRET
      }
    }
  );

  const resources = response.data.resources || [];

  return resources
    .map((resource) => {
      const designId = extractDesignId(resource.public_id || "");
      return {
        id: designId,
        name: designId,
        image: resource.secure_url || "",
        stock: 0
      };
    })
    .filter((item) => item.id && !SKIP_IDS.has(item.id))
    .sort((left, right) => left.id.localeCompare(right.id));
}

async function main() {
  if (!CLOUD_NAME || !API_KEY || !API_SECRET) {
    throw new Error(
      "CLOUDINARY_CLOUD_NAME, CLOUDINARY_API_KEY, and CLOUDINARY_API_SECRET are required."
    );
  }

  const catalog = { qualities: {} };

  for (const quality of getQualities()) {
    try {
      catalog.qualities[quality.toUpperCase()] = await fetchQualityCatalog(quality);
    } catch (error) {
      console.error(`Could not read folder: ${quality}`);
      console.error(error.response?.data || error.message);
      catalog.qualities[quality.toUpperCase()] = [];
    }
  }

  fs.mkdirSync(path.dirname(OUTPUT_FILE), { recursive: true });
  fs.writeFileSync(OUTPUT_FILE, JSON.stringify(catalog, null, 2));
  console.log(`catalog.json created successfully inside ${OUTPUT_FILE}`);
}

main().catch((error) => {
  console.error(error);
  process.exitCode = 1;
});

File: cloudinary_to_catalog.py
import json
import os

import requests
from dotenv import load_dotenv

load_dotenv()

CLOUD_NAME = os.getenv("CLOUDINARY_CLOUD_NAME", "dn1zejx9n")
API_KEY = os.getenv("CLOUDINARY_API_KEY", "154431216695392")
API_SECRET = os.getenv("CLOUDINARY_API_SECRET", "qW2aSd66mR0oMYnyBj3okM2-sE4")
BASE_FOLDER = "fabric-images"
SKIP_IDS = {"REVIEW_REQUIRED"}

catalog = {"qualities": {}}


def extract_design_id(public_id):
    filename = public_id.split("/")[-1]
    return filename.upper()


def get_qualities():
    return sorted(
        quality
        for quality in os.listdir(BASE_FOLDER)
        if os.path.isdir(os.path.join(BASE_FOLDER, quality))
    )


for quality in get_qualities():
    url = f"https://api.cloudinary.com/v1_1/{CLOUD_NAME}/resources/search"

    response = requests.post(
        url,
        auth=(API_KEY, API_SECRET),
        json={
            "expression": f"folder=fabric-images/{quality}",
            "max_results": 500,
            "sort_by": [{"public_id": "asc"}]
        }
    )

    data = response.json()

    if "resources" not in data:
        print(f"Could not read folder: {quality}")
        print(data)
        catalog["qualities"][quality.upper()] = []
        continue

    designs = []

    for resource in data["resources"]:
        public_id = resource.get("public_id", "")
        secure_url = resource.get("secure_url", "")
        design_id = extract_design_id(public_id)

        if design_id in SKIP_IDS:
            continue

        designs.append({
            "id": design_id,
            "name": design_id,
            "image": secure_url,
            "stock": 0
        })

    designs.sort(key=lambda item: item["id"])
    catalog["qualities"][quality.upper()] = designs

os.makedirs("data", exist_ok=True)

with open("data/catalog.json", "w", encoding="utf-8") as file:
    json.dump(catalog, file, indent=2)

print("catalog.json created successfully inside data/catalog.json")
